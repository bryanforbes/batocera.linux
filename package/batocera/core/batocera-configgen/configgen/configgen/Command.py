from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
