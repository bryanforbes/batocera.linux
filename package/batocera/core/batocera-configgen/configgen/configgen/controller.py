from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
