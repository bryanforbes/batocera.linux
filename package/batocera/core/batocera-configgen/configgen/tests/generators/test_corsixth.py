from __future__ import annotations

import subprocess
from pathlib import Path
from typing import TYPE_CHECKING

import pytest

from configgen.batoceraPaths import CONFIGS, ROMS
from configgen.generators.corsixth.corsixthGenerator import CorsixTHGenerator
from tests.generators.base import GeneratorBaseTest

if TYPE_CHECKING:
    from unittest.mock import Mock

    from _pytest.fixtures import SubRequest
    from pyfakefs.fake_filesystem import FakeFilesystem
    from pytest_mock import MockerFixture
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controllers
    from configgen.Emulator import Emulator


class TestCorsixTHGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[CorsixTHGenerator]:
        return CorsixTHGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'corsixth'

    @pytest.fixture
    def emulator(self) -> str:
        return 'corsixth'

    @pytest.fixture(autouse=True)
    def fs(self, fs: FakeFilesystem, request: SubRequest) -> FakeFilesystem:
        if 'no_fs_mods' not in request.keywords:
            fs.create_dir(ROMS / 'corsixth' / 'ANIMS')
            fs.create_dir(ROMS / 'corsixth' / 'DATA')
            fs.create_dir(ROMS / 'corsixth' / 'INTRO')
            fs.create_dir(ROMS / 'corsixth' / 'LEVELS')
            fs.create_dir(ROMS / 'corsixth' / 'QDATA')

        return fs

    @pytest.fixture(autouse=True)
    def subprocess_check_output(self, mocker: MockerFixture) -> Mock:
        return mocker.patch('subprocess.check_output', return_value='fr_FR')

    def test_generate(
        self,
        generator: CorsixTHGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                Path(),
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (CONFIGS / 'corsixth' / 'config.txt').read_text() == snapshot(name='config')

    def test_generate_existing(
        self,
        generator: CorsixTHGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_file(CONFIGS / 'corsixth' / 'config.txt', contents='check_for_updates = true\n')

        generator.generate(
            mock_system,
            Path(),
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'corsixth' / 'config.txt').read_text() == snapshot(name='config')

    @pytest.mark.parametrize(
        'mock_system_config',
        [
            {'cth_new_graphics': 'false'},
            {'cth_free_build_mode': 'true'},
            {'cth_play_intro': 'false'},
        ],
        ids=str,
    )
    def test_generate_config(
        self,
        generator: CorsixTHGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            Path(),
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'corsixth' / 'config.txt').read_text() == snapshot()

    @pytest.mark.parametrize(
        'lang',
        [
            'foobar',
            'en_US',
            'en_GB',
            'fr_FR',
            'oc_FR',
            'de_DE',
            'es_ES',
            'es_MX',
            'it_IT',
            'nl_NL',
            'ru_RU',
            'sv_SE',
            'cs_CZ',
            'fi_FI',
            'pl_PL',
            'hu_HU',
            'pt_PT',
            'pt_BR',
            'zh_CN',
            'zh_TW',
            'ko_KR',
            'nb_NO',
            'nn_NO',
        ],
    )
    def test_generate_language(
        self,
        generator: CorsixTHGenerator,
        lang: str,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
        subprocess_check_output: Mock,
    ) -> None:
        subprocess_check_output.return_value = lang

        generator.generate(
            mock_system,
            Path(),
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'corsixth' / 'config.txt').read_text() == snapshot()

    @pytest.mark.no_fs_mods
    def test_generate_missing_dirs_and_check_output_raises(
        self,
        generator: CorsixTHGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
        subprocess_check_output: Mock,
    ) -> None:
        subprocess_check_output.side_effect = subprocess.CalledProcessError(1, [])

        generator.generate(
            mock_system,
            Path(),
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'corsixth' / 'config.txt').read_text() == snapshot()

    def test_generate_custom_music(
        self,
        generator: CorsixTHGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_dir(ROMS / 'corsixth' / 'MP3')

        generator.generate(
            mock_system,
            Path(),
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'corsixth' / 'config.txt').read_text() == snapshot()
