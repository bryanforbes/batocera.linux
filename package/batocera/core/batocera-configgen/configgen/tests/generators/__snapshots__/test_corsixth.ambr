# serializer version: 1
# name: TestCorsixTHGenerator.test_generate
  tuple(
    list([
      'corsix-th',
      '--config-file=/userdata/system/configs/corsixth/config.txt',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
    }),
  )
# ---
# name: TestCorsixTHGenerator.test_generate[config]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[fr]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_config[{'cth_free_build_mode': 'true'}]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = true
  play_intro = true
  language = [[fr]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_config[{'cth_new_graphics': 'false'}]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = false
  free_build_mode = false
  play_intro = true
  language = [[fr]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_config[{'cth_play_intro': 'false'}]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = false
  language = [[fr]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_custom_music
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[fr]]
  audio_music = [[/userdata/roms/corsixth/MP3]]
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_existing[config]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[fr]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[cs_CZ]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[cs]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[de_DE]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[de]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[en_GB]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[en]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[en_US]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[en]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[es_ES]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[es]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[es_MX]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[es]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[fi_FI]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[fi]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[foobar]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[en]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[fr_FR]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[fr]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[hu_HU]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[hu]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[it_IT]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[it]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[ko_KR]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[ko]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[nb_NO]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[nb]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[nl_NL]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[nl]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[nn_NO]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[nb]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[oc_FR]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[fr]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[pl_PL]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[pl]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[pt_BR]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[br]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[pt_PT]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[pt]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[ru_RU]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[ru]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[sv_SE]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[sv]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[zh_CN]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[zhs]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_language[zh_TW]
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[zht]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_generate_missing_dirs_and_check_output_raises
  '''
  check_for_updates = false
  theme_hospital_install = [[/userdata/roms/corsixth]]
  unicode_font = [[/usr/share/fonts/dejavu/DejaVuSans.ttf]]
  savegames = [[/userdata/saves/corsixth]]
  screenshots = [[/userdata/screenshots]]
  fullscreen = true
  width = 1920
  height = 1080
  use_new_graphics = true
  free_build_mode = false
  play_intro = true
  language = [[en]]
  audio_music = nil
  
  '''
# ---
# name: TestCorsixTHGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
      'menu': list([
        'KEY_RIGHTSHIFT',
        'KEY_Q',
      ]),
      'reset': list([
        'KEY_RIGHTSHIFT',
        'KEY_F10',
      ]),
      'restore_state': list([
        'KEY_LEFTALT',
        'KEY_LEFTSHIFT',
        'KEY_L',
      ]),
      'save_state': list([
        'KEY_LEFTALT',
        'KEY_LEFTSHIFT',
        'KEY_S',
      ]),
    }),
    'name': 'corsixth',
  })
# ---
